src/server/zset_family.cc

```diff
@@ -1351,6 +1351,74 @@ void ZSetFamily::ZCount(CmdArgList args, ConnectionContext* cntx) {
   }
 }
 
+vector<ScoredMap> OpDiff(EngineShard* shard, Transaction* t) {
+  ArgSlice keys = t->GetShardArgs(shard->shard_id());
+  DVLOG(1) << "shard:" << shard->shard_id() << ", keys " << vector(keys.begin(), keys.end());
+  DCHECK(!keys.empty());
+
+  vector<ScoredMap> results;
+  results.reserve(keys.size());
+
+  auto& db_slice = shard->db_slice();
+  for (size_t i = 0; i < keys.size(); ++i) {
+    auto it = db_slice.Find(t->GetDbContext(), keys[i], OBJ_ZSET);
+    if (!it) {
+      results.push_back({});
```

Comment:
This is necessary, otherwise we need to do two-hops within `ZDiff`: 1. To retrieve the source set (the one we diff) 2. To retrieve the rest of the keys. Instead, if I opted in to return an empty set for the `keys` that do not exist on a given shard. That way, `ZDiff` can map and figure out the value of `source set` after it gathers the resulting sets. 

Moreover, `OpDiff` is slightly misleading, maybe I should rename it into something else, as it only `gathers` the `zsets` related to the keys.

Reply:
Agree, the name OpDiff is slightly misleading. You could rename it to OpFetch or smth else 🙂 

Yes, the single-hop approach is surely correct. This command will be mostly called with just a few arguments, so storing a non-existing default value has no overhead. Contrary, dispatching a task to another thread a second time certainly adds more latency

Reply:
@dranikpg I renamed it to `OpFetch` !

@royjacobson I like this idea as it would simplify the 3x for-loop and maybe reduce memory under certain conditions. I say maybe, because I like to think in terms of the asymptotically worst case, that is, all of the keys exist; giving us an upper bound of O(L).  My concern however is that such change would require a `union` operation which is expensive because `ScoredMap` keys would require a deep copy (they are strings) to the result set.  And then even if we don't opt in for returning a single set but rather a `vector<ScoredMap>` and yet still decide to extract the first key with a goal to not push the empty sets to the result vector, the allocator  of the vector will throw our bets about memory improvements off (because each time we push back on a vector that exceeds its capacity it has to resize and the usual allocation policy is to double at each time). 

Lastly, I am not quite sure about the thread communication part because from my understanding, we will dispatch `OpFetch` in all of the mapped shards (based on the keys used in the `zdiff` command).

Reply:
Uniting the sets makes sense when they have lots of common keys that can be cancelled out on multiple threads, instead of traversing them all in the command handler itself. If you see a simple solution you can try it out, otherwise there's no need for it